docs/dev-guide/3.0-development/rest-api.rst

```diff
@@ -0,0 +1,541 @@
+REST API Guidelines
+===================
+
+Introduction
+------------
+
+The Pulp 3 API is intended to be decoupled from the data model, which the Django REST Framework
+(DRF) makes it pretty easy to do. Where needed, support classes have been added to help the REST
+API correctly and consistently represent our specialized models and relationships (most notably
+the Master/Detail relationship).
+
+Our API starts at a :term:`DRF` :term:`Router`. Each :term:`ViewSet` is attached to this
+router, and the router's routes are exposed in urls.py. Subclasses of
+:class:`pulp.app.viewsets.NamedModelViewSet` are automatically registered with the API router,
+and most (possibly all) ViewSets created by plugins should be subclasses of this base class.
+NamedModelViewSets are associated with a Django queryset, a :term:`Serializer` that is able to
+represent members of the Django queryset in the API, and an endpoint name used when registering
+the ViewSet with the API router.
+
+All models exposed via the API must have a corresponding Serializer. Each NamedModelViewSet must
+be related must be related to a serializer and a queryset of model instances to serialize.
+Since Serializers and ViewSets are so closely related to the models they represent, the
+serializers and viewsets directories are laid out similarly to the models directory to help keep
+things consistent and easy to find.
+
+The API basic component tree looks like this::
+
+    router
+    |-- viewset
+    |   |-- queryset
+    |   |-- serializer
+    |
+    |-- viewset
+    |   |-- ...
+    |
+    |-- ...
+
+When creating API components, consider these guidelines:
+
+* Where possible, API components representing models will be defined in files whose names match
+  the corresponding file names of the models represented. For example, if you're defining the
+  serializer for a model found in pulp.app.models.consumer, the serializer should be defined in
+  pulp.app.serializers.consumer, and imported by name into pulp.app.serializers.
+
+* All objects represented in the REST API will be referred to by a single complete URL to that
+  object, using a DRF HyperlinkedRelatedField or subclass. Non-hyperlinked relations (e.g.
+  PrimaryKeyRelatedField, SlugRelatedField, etc) should be avoided. See the "Serializer
+  Relationships" section below for more details. In the database an object is identified by its
+  Primary Key. In the API an object is identified by its URL.
+
+* :class:`pulp.app.viewsets.NamedModelViewSet` subclasses defined in a plugin's "viewsets" module
+  are automatically registered with the API router. Endpoint names should follow our model naming
+  guidelines and be singular, not plural (e.g. /api/v3/repository/, not /api/v3/repositories/).
```

Comment:
I've been struggling to find a good document to back this choice up, and not doing so well.

At the moment, the best I've got to support it is that every single example that I've seen in the DRF docs uses plural endpoint names, as seen in the tutorial: http://www.django-rest-framework.org/tutorial/6-viewsets-and-routers/